modules/qualitycontrol/QualityController.ts

```diff
@@ -0,0 +1,366 @@
+            newLastN = LAST_N_THRESHOLD;
+        }
+
+        this.receiveVideoController.setLastN(newLastN);
+
+        return true;
+    }
+
+    /**
+     * Adjusts the requested resolution for remote video sources by updating the receiver constraints in an attempt to
+     * improve the encode resolution of the outbound video streams.
+     * @return {void}
+     */
+    _maybeLowerReceiveResolution() {
+        const currentConstraints = this.receiveVideoController.getCurrentReceiverConstraints();
+        const individualConstraints = currentConstraints.constraints;
+        let maxHeight = 0;
+
+        if (individualConstraints && Object.keys(individualConstraints).length) {
+            /* eslint-disable no-unused-vars */
+            for (const [key, value] of Array.from(Object.entries(individualConstraints))) {
+                const k: string = key;
+                const v: any = value;
+                maxHeight = Math.max(maxHeight, v.maxHeight);
+            }
+        }
+
+        const currentLevel = VIDEO_QUALITY_LEVELS.findIndex(lvl => lvl.height <= maxHeight);
+
+        // Do not lower the resolution to less than 180p.
+        if (currentLevel === VIDEO_QUALITY_LEVELS.length - 3) {
+            return;
+        }
+
+        this.receiveVideoController.setPreferredReceiveMaxFrameHeight(VIDEO_QUALITY_LEVELS[currentLevel + 1].height);
+    }
+
+    /**
+     * Updates the codec preference order for the local endpoint on the active media session and switches the video
+     * codec if needed.
+     *
+     * @param {string} trackId - The track ID of the local video track for which stats have been captured.
+     * @returns {boolean} - Returns true if video codec was changed.
+     */
+    _maybeSwitchVideoCodec(trackId: string): boolean {
+        const stats = this._encodeTimeStats.get(trackId);
+        const { codec, encodeResolution, localTrack } = stats.get(stats.size() - 1);
+        const codecsByVideoType = VIDEO_CODECS_BY_COMPLEXITY[localTrack.getVideoType()];
+        const codecIndex = codecsByVideoType.findIndex(val => val === codec.toLowerCase());
+
+        // Do nothing if the encoder is using the lowest complexity codec already.
+        if (codecIndex === codecsByVideoType.length - 1) {
+            return false;
+        }
+
+        if (!this._limitedByCpuTimeout) {
+            this._limitedByCpuTimeout = window.setTimeout(() => {
+                this._limitedByCpuTimeout = undefined;
+                const updatedStats = this._encodeTimeStats.get(trackId);
+                const latestSourceStats: ISourceStats = updatedStats.get(updatedStats.size() - 1);
+
+                // If the encoder is still limited by CPU, switch to a lower complexity codec.
+                if (latestSourceStats.qualityLimitationReason === QualityLimitationReason.CPU
+                    || encodeResolution <  Math.min(localTrack.maxEnabledResolution, localTrack.getCaptureResolution())) {
```

Comment:
It could but only in the cases where the camera doesn't support the requested resolution for one of the lower spatial layers when simulcast is enabled. If the requested is 720p and the camera supports only up to 540p, then we take the `Math.min(requestedSresolution, captureResolution) ` so it should be fine here. However, if the camera supports 540p and the bridge requests 360p, 270p will be sent and that should trigger a codec switch. I will improve this check to handle this case. Thanks for pointing it out.